### DISCUSSION TITLE
Use favicon as window icon in native mode

### DISCUSSION THREAD

**Discussion by falkoschindler** [zauberzeug/nicegui#1745] [Ideas / Feature Requests]
In #620, #633 and #634 we discussed ways to use the favicon argument of `ui.run` for the window icon in native mode. So far there haven't been convincing solutions. Since #620 is not a bug but a missing feature and it's currently not actively developed, I'll create a new feature request. Maybe someone has an idea how to implement it? Or maybe pywebview comes up with a API for us to use.

**Comment by FabianGoessling** [comment]
Hi, i am currently testing some possible solutions, may i ask what exactly did not convince? #634 was rejected because of the platform specific dependencies, which is understandable. But if we agree on that, how should the feature ever be included if not provided by pywebview itself (which uses platform specific deps). 
The example provided in #620 or #634 could maybe enhanced by getting the windows handle in a nicer way, but asides from that, this is how we could change the icon on the fly, similar to e.g. [pysystray](https://pywebview.flowrl.com/examples/pystray_icon.html). Happy to hear your ideas :)

**Comment by rodja** [comment]
We want to avoid maintaining a platform specific solution. So you are right: Ideally this feature is implemented in pywebview itself. An alternative could be another (well-maintained) library which provides this feature.

**Comment by FabianGoessling** [comment]
Understood, but if i recall correctly the maintainer of pywebview stated that he has no intention of adding that feature and instead pyinstallers capabilities should be used as "freezing" the app is the standard way go.

Thus the only way of ever bringing the app icon to nicegui native is finding *another* package which has a "set_icon(window_handle, icon_path)" method and is working for any OS? Chances are minimal i guess..

So if ctypes or say a *pywin32* variant are not in scope, the native mode won't have an icon out of the box soon.

**Comment by falkoschindler** [comment]
These are very valid points, @FabianGoessling. We would really like to see custom app icons in native NiceGUI apps.
I just discussed this matter with @rodja. We come to the conclusion that there's probably no better way than to include some platform-specific code.
- But in contrast to PR https://github.com/zauberzeug/nicegui/pull/634 we should improve the separation, e.g. move all icon-related code into a module native_icon.py, so that native.py only needs to import and call it.
- Furthermore we should use the `favicon` argument from `ui.run` (e.g. via the `globals` module) to avoid the need for another `ui.run` parameter.
- Since `favicon` can be all kinds of things like emojis and URLs, native_icon.py could simply raise a `NotImplementedError` in these cases for now.

What do you think? Would you like to create a pull request? 🙂

**Comment by FabianGoessling** [comment]
Makes sense, ill be happy to share my solution when it is ready. I suspect I will have to stick with ctypes as eg pywin32 would add another dep. Agree?

Points one and two are fine for me, but i think a NotImplementedError might suggest that it is raised because no one put in the work yet. But including an icon converter or something like that might not be the way to go, so i think the error needs to be something more "strict". Will look into that when i have a running solution..

**Reply by rodja** [reply]
>  I suspect I will have to stick with ctypes as eg pywin32 would add another dep. Agree?

Yes, that would be ideal.

> a NotImplementedError might suggest that it is raised because no one put in the work yet

I think it would be awesome if we integrate a emoji to bitmap converter for this feature later on... and it may not even be to hard, I hope.

**Comment by rwaal** [comment]
Hey everyone, I was just wondering if there’s been any recent activity around the topic of icons? It’s been almost a year since the last update in this thread. I’m not trying to push anything, just curious about the current status. Thanks!

**Reply by rodja** [reply]
There is no work on this feature that we maintainers are aware of. Anyone interested in giving it a try?

**Reply by FabianGoessling** [reply]
I have a working example, which is basically what was suggested in one of the previous threads. But this is a windows specific solution and feels pretty hacky (and breaks if you pin the app icon to the task bar, if i recall correctly)

If you are only looking for a way to set the icon, i can look it up, let me know..

**Reply by rwaal** [reply]
I’m not a very experienced Python developer, so I don’t think it’s realistic for me to build this functionality on my own. Not having the functionality is not super big problem, I could also look into using py2app or Nuitka.

**Reply by raphaelproux** [reply]
It seems Pywebview [implemented this](https://pywebview.flowrl.com/examples/icon.html) a couple of months ago - at least partly as it only supports Qt and GTK. Does this help?

**Reply by jdegenstein** [reply]
I would also like to know if what @raphaelproux pointed out is relevant to nicegui, thanks for the great project by the way!

**Reply by falkoschindler** [reply]
I'm not sure. You can already try setting an icon using the generic start arguments like this:
```py
from nicegui import app, ui

app.native.start_args['icon'] = 'icon.png'
ui.label('Hello world!')

ui.run(native=True)
```
It doesn't work for me though. But the reason might be that it requires GTK or QT, and I'm on macOS?

**Reply by jdegenstein** [reply]
Thank you, I tried on Windows and the above sample does not work either.

**Reply by falkoschindler** [reply]
Hm, does it work in a plain pywebview example?
```py
import webview

webview.create_window('Hello World', html='<h1>Hello, World!</h1>')
webview.start(icon='icon.png')
```

**Reply by jdegenstein** [reply]
Unfortunately no, that also does not work on Windows.

**Comment by o12486vs2** [comment]
I set the favicon in native mode with the following code
```
iconPath = "assets\\icon.png"
app.native.start_args['icon'] = os.path.abspath(iconPath)
```
However, when testing the code in native mode, the favicon wasn't applied. But after specifying an .ico file as the icon in PyInstaller, it worked!

![image](https://github.com/user-attachments/assets/f991a7ce-679f-4c9d-bf32-1b0fc0595b65)

**Comment by DearRude** [comment]
Hello. Is there any updates on this? @falkoschindler

**Reply by falkoschindler** [reply]
No relevant updates in pywebview or NiceGUI that I'm aware of. The icon parameter works, but only on GTK/Qt backends (Linux) - Windows and macOS still require setting the icon during packaging.

A platform-specific implementation using ctypes for Windows/macOS might be possible, but it's not on our roadmap at the moment. PRs are welcome though!

**Comment by falkoschindler** [comment]
Good news, everyone: thanks to @DearRude's PR #5841, native NiceGUI windows on **Windows** can now pick up the `favicon` argument at runtime (via `LoadImageW` + `WM_SETICON`). No packaging step required — just pass a file path to `ui.run(favicon=..., native=True)` and it's applied to the taskbar and title bar.

This will ship with NiceGUI 3.11.

**Reply by paco-sevilla** [reply]
FYI: [Version 6.2 of pywebview](https://github.com/r0x0r/pywebview/releases/tag/6.2) (released just a few days before #5841) allows setting the icon on Windows and Mac, as well.
See https://github.com/r0x0r/pywebview/issues/1411#issuecomment-4730832815

As mentioned above, for NiceGUI you can set `app.native.start_args['icon'] = <file_path>` before `ui.run`.
The file has to be `*.ico` on Windows and on a `*.icns` on Mac.

This way, I think the icons are shown in a much better resolution than when using the favicon...
